packages/expo-image/src/ExpoImage.web.tsx

```diff
@@ -57,50 +62,149 @@ function getObjectPositionFromContentPosition(contentPosition?: ImageContentPosi
     .join(' ');
 }
 
-const ensureIsArray = <T extends any>(source: T | T[] | undefined) => {
+function ensureIsArray(source?: ImageSource): (ImageUriSource | RequireSource)[] {
   if (Array.isArray(source)) {
     return source;
   }
   if (source == null) {
     return [];
   }
   return [source];
+}
+
+type ImageState = 'empty' | 'loading' | 'loaded' | 'error';
+function useImageState(source: ImageSource | undefined) {
+  const [imageState, setImageState] = React.useState<ImageState>(source ? 'loading' : 'empty');
+  React.useEffect(() => {
+    setImageState((prevState) =>
+      prevState === 'empty' ? (source ? 'loading' : 'empty') : prevState
+    );
+  }, [source]);
+
+  const onLoad = React.useCallback(
+    () => setImageState((prevState) => (imageState === 'loading' ? 'loaded' : prevState)),
+    []
+  );
+  const handlers = React.useMemo(
+    () => ({
+      onLoad,
+    }),
+    [onLoad]
+  );
+  return [imageState, handlers] as [ImageState, { onLoad: () => void }];
+}
+
+const getCSSTiming = (timing?: ImageTransitionTiming) => {
+  return {
+    [ImageTransitionTiming.EASE_IN]: 'ease-in',
+    [ImageTransitionTiming.EASE_OUT]: 'ease-out',
+    [ImageTransitionTiming.EASE_IN_OUT]: 'ease-in-out',
+    [ImageTransitionTiming.LINEAR]: 'linear',
+  }[timing || ImageTransitionTiming.LINEAR];
 };
 
+function useTransition(
+  transition: ImageTransition | null | undefined,
+  state: ImageState
+): Record<'placeholder' | 'image', Partial<React.CSSProperties>> {
+  if (!transition?.effect) return { placeholder: {}, image: {} };
+  const { duration, timing, effect } = {
+    timing: ImageTransitionTiming.EASE_IN_OUT,
+    duration: 1000,
+    ...transition,
+  };
+  if (effect === ImageTransitionEffect.CROSS_DISOLVE) {
+    const commonStyles = {
+      transition: `opacity ${duration}ms`,
+      transitionTimingFunction: getCSSTiming(timing),
+    };
+    return {
+      image: {
+        opacity: state === 'loaded' ? '1' : '0',
+        ...commonStyles,
+      },
+      placeholder: {
+        opacity: state === 'loaded' ? '0' : '1',
+        ...commonStyles,
+      },
+    };
+  } else if (effect === ImageTransitionEffect.FLIP_FROM_TOP) {
```

Comment:
I know I'm nitpicking 😅 but since the previous `if` always returns I think it would look better to skip `else` 😉
```suggestion
  }
  if (effect === ImageTransitionEffect.FLIP_FROM_TOP) {
```

Reply:
nah, that's a fair point!
Missed that since I assumed we'd make that a switch case anyways when we add more styles